tasks/libs/pipeline/tools.py

```diff
@@ -77,13 +77,29 @@ def gracefully_cancel_pipeline(repo: Project, pipeline: ProjectPipeline, force_c
     """
 
     jobs = pipeline.jobs.list(per_page=100, all=True)
+    kmt_cleanup_jobs_to_run: set[str] = set()
 
     for job in jobs:
         if job.stage in force_cancel_stages or (
-            job.status not in ["running", "canceled"] and "cleanup" not in job.name
+            job.status not in ["running", "canceled", "success"] and "cleanup" not in job.name
         ):
             repo.jobs.get(job.id, lazy=True).cancel()
 
+            if job.name.startswith("kmt_setup_env") or job.name.startswith("kmt_run"):
+                component = "sysprobe" if "sysprobe" in job.name else "secagent"
+                arch = "x64" if "x64" in job.name else "arm64"
+                cleanup_job = f"kmt_{component}_cleanup_{arch}_manual"
+                kmt_cleanup_jobs_to_run.add(cleanup_job)
+
+    # Run manual cleanup jobs for KMT. If we canceled the setup env or the tests job,
+    # the cleanup job will not run automatically. We need to trigger the manual variants
+    # to avoid leaving KMT instances running too much time.
+    for cleanup_job in kmt_cleanup_jobs_to_run:
+        for job in jobs:
+            if job.name == cleanup_job:
+                print(f"Triggering KMT {cleanup_job} job")
+                repo.jobs.get(job.id, lazy=True).play()
```

Comment:
nit - but just a thought because it took me a bit to get that you had to loop over again because you were looking for the `job.id`. I think you could create a dict of `name`->`id` in the first loop and just use it to avoid the second loop.

Also I think you can `break` in that `if` if you want to keep it that way 🙂

Reply:
Changed, thx!